### ISSUE TITLE
v7.6.6: Backward RegEx Search does not work for UTF-8 Characters or Strings in a [Normal] Text File

### ISSUE THREAD

**Issue by Doryforos** [notepad-plus-plus/notepad-plus-plus#5515]
Backward **RegEx** Search [Normal & Extended Search Modes OK], either SHIFT+F3, or SHIFT+ENTER or SHIFT+CLICK when in the FIND Dialogue, or menu, does not work, either for UTF-8 or for non-Latin ANSI characters or character strings in a [Normal] Text File — it only works for single-byte characters or strings; if the character or string is followed by a space, though, backward search works OK, but when a space precedes the character or string it does not work.

Earlier versions of Notepad++ (e.g. v6.2.2, when run from a USB flash disk plugged on the same computer and OS) worked (and still work) fine in this regard.

Fairly easy to reproduce:
Get some substantial [3 pages or more, in order to have multiple occurrences of the same word(s)] amount of text in Greek, from, e.g., [a Greek newspaper](http://www.kathimerini.gr/), in a buffer and save it in either UTF-8 or ANSI [ISO-8859-7 or Windows-1253 — or do not save it, neither matters], **set the Search Mode to “Re&gular expression”** and perform a backward search for either a litteral character or a word or string, or a regular expression, in the Greek alphabet: Npp will only find forwards, but if the same is attempted by adding a space at the end of the search string, it will also find backwards.

Possibly Related: TextFX Characters cannot hard wrap Greek text at specified column, it wraps it **at about half of what has been specified** — go figure!

My Setup [a fresh, clean installation, with no carry-overs from my previous, honourably retired, XP installation]:
Notepad++ v7.6.6 on MS-Win7 Ultimate + SP 1 + Updates

My Debug Info:

a. Installed 32bit [I **need** TextFX Characters]:
```
Notepad++ v7.6.6   (32-bit)
Build time : Apr  3 2019 - 23:49:50
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 7 (64-bit)
Plugins : Explorer.dll HexEditor.dll LightExplorer.dll mimeTools.dll NppConverter.dll NppExec.dll NppExport.dll NppTextFX.dll
```

b. Portable 64bit, unzipped and run without any extra plugins or any customization whatsoever:
```
Notepad++ v7.6.6   (64-bit)
Build time : Apr  3 2019 - 23:52:32
Path : [UFD]:\Software\npp766\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS : Windows 7 (64-bit)
Plugins : mimeTools.dll NppConverter.dll
```

The bug reproduces consistently in both of the above, also in previous versions, **since I first reported this back in July 2015**.

Related Issue:
#534 v6.8: Backward RegEx Search does not work for UTF-8 Characters in a [Normal] Text File

**Comment by Ekopalypse** [comment]
By changing [cpMax to cpMin](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp#L1546) it seems to work. 
Sample text
€ sign for an euro, 
so let's find the € 
in this text of
€s 
and was looking for € of course.
Worked in normal and RegEx mode.

**Comment by MetaChuh** [comment]
verified and reproducible.
(see screencast below)

**steps to reproduce:**

* open up the attached test file:
[#5515 Doryforos, v7.6.6 Backward RegEx Search does not work for UTF-8 Characters or Strings in a [Normal] Text File.txt](https://github.com/notepad-plus-plus/notepad-plus-plus/files/3068221/5515.Doryforos.v7.6.6.Backward.RegEx.Search.does.not.work.for.UTF-8.Characters.or.Strings.in.a.Normal.Text.File.txt)
* search for `στα` in regex mode.
   * forward search: ok
   * **backward search: no function**

* working: search for `στα ` in regex mode. (added space at the end of στα)
   * forward search: ok
   * backward search: ok

![#5515 Doryforos, v7 6 6 Backward RegEx Search does not work for UTF-8 Characters or Strings in a  Normal  Text File](https://user-images.githubusercontent.com/45773305/55951662-1292d180-5c58-11e9-80e1-70bdf61572df.gif)

debug info:
<kbd>Notepad++ v7.6.6   (64-bit)
Build time : Apr  3 2019 - 23:52:32
Path : C:\Users\User\Desktop\npp.7.6.6.bin.x64\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS : Windows 7 (64-bit)
Plugins : mimeTools.dll NppConverter.dll</kbd>

<kbd>Notepad++ v7.6.6   (32-bit)
Build time : Apr  3 2019 - 23:49:50
Path : C:\Users\User\Desktop\npp.7.6.6.bin\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS : Windows 7 (64-bit)
Plugins : mimeTools.dll NppConverter.dll NppExport.dll</kbd>

**Comment by sasumner** [comment]
How come everyone always ignores the fact that Notepad++ is "undecided" about whether backward regex search is allowable?  Note in screencast above that "Backward direction" checkbox is greyed/disabled.  It sure seems like the undecided nature of that should be settled (and made 100% consistent) once and for all before deciding upon this current issue.

**Comment by Ekopalypse** [comment]
@sasumner 
> Notepad++ is "undecided" about whether backward regex search is allowable

May I ask you where this comes from?
I saw, that the backward checkbox is disabled if one checks regular expression,
but never thought that this is done intentionally.

**Comment by sasumner** [comment]
@Ekopalypse 

>I saw, that the backward checkbox is disabled...but never thought that this is done intentionally

Hmmm...I looked through all of the issues you have reported and don't see a bug report from you on this! :)

**Comment by Ekopalypse** [comment]
@sasumner - simply because I didn't care about it as I wasn't affected at all.

**Comment by sasumner** [comment]
@Ekopalypse So I'll guess about the history:  If you are going to do a forward search, you just hand over a buffer and a regex to a regex-engine and it will give you the matching position. So, say you are searching forward for a regex that will be found after 100 characters from the caret position.  You (N++ coder) make one call to the engine and you are done (you get told the match is 100 characters in). If you are going to search backward, it gets trickier.  You have to go backward one character at a time, and feed the engine that buffer each time until a match is found. So, for a similar example, if you are searching backward and your match will happen 100 characters before the caret position, you (N++ coder) have to make 100 calls to the regex engine, and you (N++ coder) have to administrate that (but that is true for non-regex search so this argument is weak).  Plus, if the regex is complicated, all the failing searches could add up to a lot of time where the user is wondering what is going on.  So I might hazard to guess that Mr. N++ coder initially thought backward regex searches were a "bad idea" and they were disallowed.  But then some waffling crept in, and they were implicitly allowed, and then the `<<` button was added, and they were still fine, but meanwhile the "Backward direction" button logic never followed along...so it is disabled to present day when search mode is set to regex.  This is all just a guess; I'm too lazy to look at the source.  I could be totally wrong and its inconsistent behavior has its roots in something else.

**Comment by MetaChuh** [comment]
@sasumner @Ekopalypse 

is there any regex workaround you can provide to the author ?

**Comment by Ekopalypse** [comment]
@sasumner - I've tested a couple of regexes (lookahead and lookbehind included) and all seemed to work,
but I must admit I'm not a regex expert at all. 
I was searching the code and I've found a [comment](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/4c548d187d1c83ab887e5411d1bb4e8ae52b1bd7/PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp#L330) that there is (maybe was) a problem since 6.3 but was not able to find out 
what the issue is/was. In addition, I haven't really understood what the boost regex engine is doing
but from npps point of view it is transparent. Npp just provides the start and end position which either is 
current_position to 0 (backward) or current_position to text_length (forward).

@MetaChuh - you asking me?? (hip-hop-slang) and confused_emoji  :-)
I'm just about learning those nasty regex things

**Comment by MetaChuh** [comment]
@Ekopalypse 
> the next opening doesn't disable the backward checkbox.

yes.

@ both:
we have our hangout place over at the community, please avoid any dialogues that could lead to ftldr at the tracker.

**Comment by sasumner** [comment]
@Ekopalypse So my thoughts were for nothing. :)  But I still think (there I go again, doing the dangerous) that asking a search algo that loves to process from left to right to go right to left can get into some time-expensive situations.  As far as the "comment" you linked to in the code, I guess we may never know the origins.

**Comment by sasumner** [comment]
Another documented example where forward-replacement works and backward-replacement doesn't:  https://notepad-plus-plus.org/community/topic/18102/regex-partial-replacement-syntax-grouping-not-working

There is something wacky about regular expressions and moving backward through a document.  I think the code (that is halfway there) to prevent the user from doing backward searches knew something that we have forgotten.

**Comment by sasumner** [comment]
So I've learned some things since my comment that starts out "So I'll guess about the history" above, in conjunction with my research on #3640 .

What I've learned from rooting thru the code (and discussions with @Ekopalypse in emails) is that when doing a backward regex search, the code will start at position zero in a document (no matter where the caret is positioned) and will search from there (zero) looking at for matches.  It will remember a match and then keep looking for a "better" (closer to the caret) match.  It will repeat this as many times as necessary to get to the match closest to (but still before) the caret position; it will then dub this as the result of the backward regex search.

So, if you are in a document filled with one million `t` characters (only) and you are at the position of the millionth `t` and you request a backward regex search for `t`, the code will find one million previous matches before finding the match it hands back to the user.

The relevant code can be found in the `FindTextBackward` function of `BoostRegExSearch.cxx`, specifically here: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/scintilla/boostregex/BoostRegExSearch.cxx#L373
